flink-runtime/src/main/java/org/apache/flink/runtime/failure/FailureEnricherUtils.java

```diff
@@ -0,0 +1,227 @@
+                    + " because Flink shouldn't allow registering enrichers with the same output.";
+
+    public static Collection<FailureEnricher> getFailureEnrichers(
+            final Configuration configuration) {
+        final PluginManager pluginManager =
+                PluginUtils.createPluginManagerFromRootFolder(configuration);
+        return getFailureEnrichers(configuration, pluginManager);
+    }
+
+    /**
+     * Returns a set of validated FailureEnrichers for a given configuration, job ID, job name,
+     * metric group, and plugin manager.
+     *
+     * @param configuration the configuration for the job
+     * @param pluginManager the PluginManager for the job
+     * @return a collection of validated FailureEnrichers
+     * @throws IllegalArgumentException if any of the FailureEnrichers are invalid
+     */
+    static Collection<FailureEnricher> getFailureEnrichers(
+            final Configuration configuration, final PluginManager pluginManager) {
+        Set<String> includedEnrichers = getIncludedFailureEnrichers(configuration);
+        //  When empty, NO enrichers will be started.
+        if (includedEnrichers.isEmpty()) {
+            return Collections.emptySet();
+        }
+        final Iterator<FailureEnricherFactory> factoryIterator =
+                pluginManager.load(FailureEnricherFactory.class);
+        final Set<FailureEnricher> failureEnrichers = new HashSet<>();
+        while (factoryIterator.hasNext()) {
+            try {
+                final FailureEnricherFactory failureEnricherFactory = factoryIterator.next();
+                final FailureEnricher failureEnricher =
+                        failureEnricherFactory.createFailureEnricher(configuration);
+                if (includedEnrichers.contains(failureEnricher.getClass().getName())) {
+                    failureEnrichers.add(failureEnricher);
+                    LOG.debug(
+                            "Found failure enricher {} at {} ",
+                            failureEnricherFactory.getClass().getName(),
+                            new File(
+                                            failureEnricher
+                                                    .getClass()
+                                                    .getProtectionDomain()
+                                                    .getCodeSource()
+                                                    .getLocation()
+                                                    .toURI())
+                                    .getCanonicalPath());
+                } else {
+                    LOG.info(
+                            "Excluding failure enricher {}, not configured in enricher list ({}).",
+                            failureEnricherFactory.getClass().getName(),
+                            includedEnrichers);
+                }
+            } catch (Exception e) {
+                LOG.warn("Error while loading failure enricher factory.", e);
```

Comment:
I'd personally throw an exception here because this violates the cluster setup that the user has asked for. Also, since this happens during JM startup, it should be fine to fail fast.

Reply:
I agree, but lets leave it as is for now and add the exception when we decide where we are call the factory (JM/Dispatcher/ClusterEntryPoint)